graphql/src/Mu/GraphQL/Quasi.hs

```diff
@@ -0,0 +1,166 @@
+      [t| 'OptionalRef ('ListRef $(gtypeToType a)) |]
+    gtypeToType _ = fail "this should not happen, please, file an issue"
+typeToDec _ _ (GQL.TypeDefinitionInterface _)       = fail "interface types are not supported"
+typeToDec _ _ (GQL.TypeDefinitionUnion _)           = fail "union types are not supported"
+typeToDec _ _ (GQL.TypeDefinitionEnum enums)        = enumToDecl enums
+  where
+    enumToDecl :: GQL.EnumTypeDefinition -> Q Result
+    enumToDecl (GQL.EnumTypeDefinition _ (coerce -> name) _ symbols) =
+      GQLSchema <$> [t|'DEnum $(textToStrLit name)
+                              $(typesToList <$> traverse gqlChoiceToType symbols)|]
+    gqlChoiceToType :: GQL.EnumValueDefinition -> Q Type
+    gqlChoiceToType (GQL.EnumValueDefinition _ (coerce -> c) _) =
+      [t|'ChoiceDef $(textToStrLit c)|]
+typeToDec _ _ (GQL.TypeDefinitionInputObject inpts) = inputObjToDec inpts
+  where
+    inputObjToDec :: GQL.InputObjectTypeDefinition -> Q Result
+    inputObjToDec (GQL.InputObjectTypeDefinition _ (coerce -> name) _ fields) =
+        GQLSchema <$> [t|'DRecord $(textToStrLit name)
+                                  $(typesToList <$> traverse gqlFieldToType fields)|]
+    gqlFieldToType :: GQL.InputValueDefinition -> Q Type
+    gqlFieldToType (GQL.InputValueDefinition _ (coerce -> fname) ftype _) =
+      [t|'FieldDef $(textToStrLit fname) $(ginputTypeToType ftype)|]
+    ginputTypeToType :: GQL.GType -> Q Type
+    ginputTypeToType (GQL.TypeNamed (coerce -> False) (coerce -> a)) =
+      [t| $(typeToPrimType a) |]
+    ginputTypeToType (GQL.TypeNamed (coerce -> True) (coerce -> a)) =
+      [t| 'OptionalRef $(typeToPrimType a) |]
+    ginputTypeToType (GQL.TypeList (coerce -> False) (coerce -> a)) =
+      [t| 'ListRef $(ginputTypeToType a) |]
+    ginputTypeToType (GQL.TypeList (coerce -> True) (coerce -> a)) =
+      [t| 'OptionalRef ('ListRef $(ginputTypeToType a)) |]
+    ginputTypeToType _ = fail "this should not happen, please, file an issue"
+    typeToPrimType :: GQL.Name -> Q Type
+    typeToPrimType (GQL.unName -> "Int")     = [t|'TPrimitive Integer|]
+    typeToPrimType (GQL.unName -> "Float")   = [t|'TPrimitive Double|]
+    typeToPrimType (GQL.unName -> "String")  = [t|'TPrimitive T.Text|]
+    typeToPrimType (GQL.unName -> "Boolean") = [t|'TPrimitive Bool|]
+    typeToPrimType (GQL.unName -> "ID")      = [t|'TPrimitive UUID|]
+    typeToPrimType (coerce -> name)          = [t|'TSchematic $(textToStrLit name)|]
+
+scalarToType :: GQL.Name -> TypeMap -> Name -> Q Type
+scalarToType (GQL.unName -> "Int") _  _     = [t|'PrimitiveRef Integer|]
+scalarToType (GQL.unName -> "Float") _ _    = [t|'PrimitiveRef Double|]
+scalarToType (GQL.unName -> "String") _ _   = [t|'PrimitiveRef T.Text|]
+scalarToType (GQL.unName -> "Boolean") _ _  = [t|'PrimitiveRef Bool|]
+scalarToType (GQL.unName -> "ID") _ _       = [t|'PrimitiveRef UUID|]
+scalarToType (coerce -> name) tm schemaName =
```

Comment:
The name `scalarToType` suggests that if the type is not a scalar you should fail instead. Maybe rename the function?